cli/admin/src/test/java/io/pravega/cli/admin/bookkeeper/BookkeeperCommandsTest.java

```diff
@@ -0,0 +1,191 @@
+        Assert.assertTrue(outContent.toString().contains("Deleted Ledger 0"));
+        command.deleteCandidates(Collections.singletonList(-1L), Collections.singletonList(0L), context);
+        command.deleteCandidates(Collections.singletonList(0L), Collections.singletonList(1L), null);
+    }
+
+    @Test
+    public void testBookKeeperRecoveryCommand() throws Exception {
+        createLedgerInBookkeeperTestCluster(0);
+        String commandResult = TestUtils.executeCommand("container recover 0", STATE.get());
+        Assert.assertTrue(commandResult.contains("Recovery complete"));
+        CommandArgs args = new CommandArgs(Collections.singletonList("0"), STATE.get());
+        ContainerRecoverCommand command = new ContainerRecoverCommand(args);
+        // Test unwrap exception options.
+        command.unwrapDataCorruptionException(new DataCorruptionException("test"));
+        command.unwrapDataCorruptionException(new DataCorruptionException("test", "test"));
+        command.unwrapDataCorruptionException(new DataCorruptionException("test", Arrays.asList("test", "test")));
+        // Check that exception is thrown if ZK is not available.
+        this.zkUtil.stopCluster();
+        AssertExtensions.assertThrows(DataLogNotAvailableException.class, () -> TestUtils.executeCommand("container recover 0", STATE.get()));
+    }
+
+    @Test
+    public void testRecoveryState() {
+        CommandArgs args = new CommandArgs(Collections.singletonList("0"), STATE.get());
+        ContainerRecoverCommand.RecoveryState state = new ContainerRecoverCommand(args).new RecoveryState();
+        Operation op = new TestOperation();
+        List<DataFrameRecord.EntryInfo> entries = new ArrayList<>();
+        entries.add(new TestEntryInfo());
+        // Exercise RecoveryState logic.
+        state.operationComplete(op, new DataCorruptionException("Test exception"));
+        state.newOperation(op, entries);
+        state.operationComplete(op, null);
+    }
+
+    private void createLedgerInBookkeeperTestCluster(int logId) throws Exception {
+        BookKeeperConfig bookKeeperConfig = BookKeeperConfig.builder().with(BookKeeperConfig.ZK_METADATA_PATH, "ledgers")
+                .with(BookKeeperConfig.BK_LEDGER_PATH, "/ledgers")
+                .with(BookKeeperConfig.ZK_ADDRESS, zkUtil.getZooKeeperConnectString()).build();
+        @Cleanup
+        CuratorFramework curatorFramework = CuratorFrameworkFactory.newClient(zkUtil.getZooKeeperConnectString(), new RetryOneTime(5000));
+        curatorFramework.start();
+        @Cleanup
+        BookKeeperLogFactory bookKeeperLogFactory = new BookKeeperLogFactory(bookKeeperConfig, curatorFramework, Executors.newSingleThreadScheduledExecutor());
+        bookKeeperLogFactory.initialize();
+        @Cleanup
+        DurableDataLog log = bookKeeperLogFactory.createDurableDataLog(logId);
+        log.initialize(Duration.ofSeconds(5));
+    }
+
+    private static class TestOperation extends Operation {
```

Comment:
Is it required?

Reply:
`Operation` is an abstract class, so it cannot be instantiated. So, either we use any non-abstract class extending `Operation` or we create a mock one for testing purposes as we do here.